### DISCUSSION TITLE
CreateAction and duplicate entry

### DISCUSSION THREAD

**Discussion by photojo** [filamentphp/filament#9769] [Help] [answered]
I'm using CreateAction on filament 2

The **CreateAction** has a large form, that takes one minute to fill. One field is a hidden field including the invoice number. This number is generated by my own function. 
But, it can happen, that two DB-rows get the same invoice number, when two or more users fill in and submit the form.

id          invoice
23            856
24            856

What is the best way to avoid this? Making an **Action** with additional routine instead of CreateAction?

Thank you very very much!

**Comment by leandrocfe** [comment]
You could use hooks to validate the data before

```php
CreateAction::make()
    ->beforeFormValidated(function () {
        // Runs before the form fields are validated when the form is submitted.
    })
    ->afterFormValidated(function () {
        // Runs after the form fields are validated when the form is submitted.
    })
    ->before(function () {
        // Runs before the form fields are saved to the database.
    })
```
https://filamentphp.com/docs/2.x/admin/resources/creating-records#lifecycle-hooks

**Comment by photojo** [comment]
Thank you very much!

**Reply by photojo** [reply]
Can you help me once again, please?

 ```
 protected function getTableHeaderActions(): array
    {
        return [
            CreateAction::make('create')
                ->form([
                    TextInput::make('name')
                        ->label('name')
                        ->required()
                        ->columnSpan(2),
                   
                ])
                ->before(function (Action $action, array $data) {
                   
                // HERE I WANT TO MAKE name = 'Antony'       (*)

                }),
        ];
    }
```

This is just an simple example. Of course, I normaly use default!!!
So, how can I change the name-attribute in line (*) and save it to DB?

Thank in advance!

**Reply by leandrocfe** [reply]
`$data['name'] = 'Antony';` ?

If you want to add a default value, you could do this

```php
Actions\Action::make('create')
    ->mountUsing(fn (Form $form) => $form->fill())
    ->form([
        TextInput::make('name')
            ->label('name')
            ->required()
            ->columnSpan(2)
            ->default('Antony'),

    ])->action(function (array $data) {
        //$data['name'];
    }),
```

**Comment by photojo** [comment]
No, I have to call a function, which returns a unique invoice number, that should be saved in db.

**Reply by leandrocfe** [reply]
share some code you are trying to do

**Comment by photojo** [comment]
So, thak you very much. I got it. This works for me!
          

```
                 CreateAction::make('new')
                  ->form([
                    TextInput::make('invoice')
                            
                  ])
                ->before(function (Mymodel $mymodel) {
                    $mymodel->name = $this->invoice();
                  
                }),
```

So, it works. Before data is saved to DB my $this->invoice() returns a unique number. Fine!

BUT: Does this really prevents, that each dataset has a different invoicenumber, even if 20 users create datasets at the same time?

**Comment by photojo** [comment]
So, I'm so sorry. Obviously I made a mistake. Obove code from my last posting does not work.
I think I can not use CreateAction, but Action.

I try to describe the problem again. I have a form with about 50 fields. Just before writing data into DB I have to check DB, if a certain invoice-number already exists. How can I achieve this?

What' wrong with 


```
  Action::make('create')->mountUsing(fn(Form $form) => $form->fill(               
                ))
                ->form([
                    TextInput::make('client')  ])
                ->before(function (Bill $bill, array $data) {
                    $bill->invoice = $this->chekcInvoice();   // return unique number after DB-query
                })
                ->action(function (Bill $bill, array $data) {
                }),
```

All fields should be filled by the form-inputs, just invoice-field should be filled by  $this->chekcInvoice();

**Comment by photojo** [comment]
Now I use 
  ```
->before(function (Bill $bill, array $data) {
                    $bill->number = $this->chekcInvoice() ;
                    $bill->save();
                })
```
But in my form I also have 
```
      Repeater::make('product')
           ->label('Product')
           ->relationship()
           ->schema([ ... ])
```
to add products to the invoice.


How can I handle this inside ->before ..., because this is another model?!


So, I have to change only ONE FIELD in "->before()"

**Comment by photojo** [comment, marked as answer]
I solved it in this way:

```
  ->mutateFormDataUsing(function (array $data): array {
                    $data['invoicenumber'] = $this->makeUniqueInVoiceNumber();
                    return $data;
                }),
```

Hope, that helps other users!
